### ISSUE TITLE
Can't calculate shap values of GPU-trained model

### ISSUE THREAD

**Issue by wowone** [catboost/catboost#1284]
Problem: Given a ranking model trained on GPU with YetiRankPairwise loss_function it's not possible to calculate shap values. Switching task type to CPU resolves the issue.

Here's my minimal example based on your ranking tutorial data:
```(python)
import shap
from catboost import CatBoost, Pool
from catboost.datasets import msrank_10k
train_df, test_df = msrank_10k()

X_train = train_df.drop([0, 1], axis=1).values
y_train = train_df[0].values
queries_train = train_df[1].values

train = Pool(
    data=X_train,
    label=y_train,
    group_id=queries_train
)

model = CatBoost()
model.set_params(iterations=100,
                 loss_function="YetiRankPairwise",
                 task_type="GPU",
                 verbose=False)
model.fit(train)

explainer = shap.TreeExplainer(model)
single_item_pool = Pool(data=[X_train[0]])
shap_values = explainer.shap_values(single_item_pool)
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Metric NDCG:type=Base is not implemented on GPU. Will use CPU for metric computation, this could significantly affect learning time
Setting feature_perturbation = "tree_path_dependent" because no background data was given.
---------------------------------------------------------------------------
CatBoostError                             Traceback (most recent call last)
<ipython-input-2-a1cf0c800ada> in <module>
     23 explainer = shap.TreeExplainer(model)
     24 single_item_pool = Pool(data=[X_train[0]])
---> 25 shap_values = explainer.shap_values(single_item_pool)

~/myenv/lib/python3.6/site-packages/shap/explainers/tree.py in shap_values(self, X, y, tree_limit, approximate, check_additivity)
    262                 if type(X) != catboost.Pool:
    263                     X = catboost.Pool(X, cat_features=self.model.cat_feature_indices)
--> 264                 phi = self.model.original_model.get_feature_importance(data=X, fstr_type='ShapValues')
    265 
    266             # note we pull off the last column and keep it as our expected_value

~/catboost/catboost/python-package/catboost/core.py in get_feature_importance(self, data, type, prettified, thread_count, verbose, fstr_type, shap_mode, interaction_indices, shap_calc_type)
   2350         with log_fixup():
   2351             fstr, feature_names = self._calc_fstr(type, data, thread_count, verbose, shap_mode, interaction_indices,
-> 2352                                                   shap_calc_type)
   2353         if type in (EFstrType.PredictionValuesChange, EFstrType.LossFunctionChange, EFstrType.PredictionDiff):
   2354             feature_importances = [value[0] for value in fstr]

~/catboost/catboost/python-package/catboost/core.py in _calc_fstr(self, type, pool, thread_count, verbose, shap_mode, interaction_indices, shap_calc_type)
   1271     def _calc_fstr(self, type, pool, thread_count, verbose, shap_mode, interaction_indices, shap_calc_type):
   1272         return self._object._calc_fstr(type.name, pool, thread_count, verbose, shap_mode, interaction_indices,
-> 1273                                        shap_calc_type)
   1274 
   1275     def _calc_ostr(self, train_pool, test_pool, top_size, ostr_type, update_method, importance_values_sign, thread_count, verbose):

_catboost.pyx in _catboost._CatBoost._calc_fstr()

_catboost.pyx in _catboost._CatBoost._calc_fstr()

CatBoostError: catboost/libs/fstr/util.cpp:102: Cannot calc shap values, model contains non zero approx for zero-weight leaf

```

catboost version: 0.23
Operating System: Ubuntu 18.04
CPU: AMD Ryzen Threadripper 1920X
GPU: GeForce RTX 2070

**Comment by annaveronika** [comment]
We have reproduced the error and are working on the fix.

**Comment by bbinwang** [comment]
> We have reproduced the error and are working on the fix.

hi，what is the progress of implementation？

**Comment by shushuK** [comment]
May I know that any progress of the bug? Thanks!

**Comment by firedie** [comment]
Hi, is there any progress for this issus? I am facing a same one. Thanks!

**Comment by kizill** [comment]
Released in 0.24.4, please try it :)
